File: src/nstreamcom/nencode.py
from typing import Optional
from .sizing import DATA_BITS, as_transmission_size, as_data_size, NSIZE_SIZE, ENCODED_NSIZE_SIZE
from .bits import as_byte


def encode(data: bytes | bytearray | list[int]) -> bytearray:
    encoded: bytearray = bytearray()
    left_shift: int
    previous_stuffed: int
    ratio: int = 0

    for i in range(as_transmission_size(len(data))):
        left_shift = i % 8
        previous_stuffed = 8 - left_shift

        if left_shift == 0:
            encoded.append(as_byte(data[ratio]))
        else:
            encoded.append(as_byte(data[ratio] >> previous_stuffed))
            ratio += 1
            if ratio < len(data):
                encoded[i] |= as_byte(data[ratio] << left_shift)

        encoded[i] &= as_byte(~(0xFF << DATA_BITS))

    return encoded


def decode(encoded: bytes | bytearray | list[int]) -> bytearray:
    data: bytearray = bytearray()
    ratio: int = 0
    left_next: int
    for i in range(as_data_size(len(encoded))):
        right_shift = i % DATA_BITS
        left_next = DATA_BITS - right_shift

        if i != 0 and right_shift == 0:
            ratio += 1

        next_data: int = (encoded[ratio] & ~(0xFF << DATA_BITS)) >> right_shift
        data.append(as_byte(next_data))
        ratio += 1

        if ratio < len(encoded):
            data[i] |= as_byte(((encoded[ratio] & ~(0xFF << DATA_BITS)) << left_next))

    return data


def encode_size(data_size: int, byteorder: Optional[str] = None) -> bytearray:
    encoded: bytearray = encode(data_size.to_bytes(NSIZE_SIZE, byteorder or 'little'))
    for i in range(ENCODED_NSIZE_SIZE):
        encoded[i] |= 1 << 7
    return encoded


def encode_with_size(data: bytes | bytearray | list[int], size_byteorder: Optional[str] = None) -> bytearray:
    encoded: bytearray = encode_size(len(data), size_byteorder)
    encoded.extend(encode(data))
    return encoded


def decode_size(encoded_size: bytearray | bytes | list[int], byteorder: Optional[str] = None) -> int:
    zeroed_out: bytearray = bytearray()
    for byte in encoded_size:
        zeroed_out.append(byte & ~(1 << 7))
    return int.from_bytes(decode(zeroed_out), byteorder or 'little')

File: src/nencode.zig
const asBytes = @import("std").mem.asBytes;
const sizing = @import("sizing.zig");
const nsize_int = sizing.nsize_int;
const encoded_nsize_int = sizing.encoded_nsize_int;
const nsize_int_byte_count = sizing.nsize_int_byte_count;
const encoded_nsize_int_byte_count = sizing.encoded_nsize_int_byte_count;

pub fn encode(data: []const u8, encoded: []u8) void {
    var di: u32 = 0;
    var ei: u32 = 0;
    var left_shift: u8 = 7;
    var previous_stuffed: u8 = undefined;

    while (di < data.len and ei < encoded.len) : (ei += 1) {
        left_shift = if (left_shift == 7) 0 else left_shift + 1;
        previous_stuffed = 8 - left_shift;

        if (left_shift == 0) {
            encoded[ei] = data[di];
        } else {
            encoded[ei] = data[di] >> @intCast(previous_stuffed);
            di += 1;

            if (di < data.len) {
                encoded[ei] |= data[di] << @intCast(left_shift);
            }
        }

        encoded[ei] &= ~(@as(u8, 0xFF) << 7);
    }
}

pub fn decode(encoded: []const u8, decoded: []u8) void {
    var ei: u32 = 0;
    var di: u32 = 0;
    var right_shift: u8 = 6;
    var left_next: u8 = undefined;

    while (ei < encoded.len and di < decoded.len) : (di += 1) {
        right_shift = if (right_shift == 6) 0 else right_shift + 1;
        left_next = 7 - right_shift;

        if (di != 0 and right_shift == 0) {
            ei += 1;
            if (ei == encoded.len) {
                break;
            }
        }

        decoded[di] = (encoded[ei] & ~(@as(u8, 0xFF) << 7)) >> @intCast(right_shift);
        ei += 1;
        if (ei < encoded.len) {
            decoded[di] |= (encoded[ei] & ~(@as(u8, 0xFF) << 7)) << @intCast(left_next);
        }
    }
}

pub fn encodeSize(size: nsize_int) encoded_nsize_int {
    var encoded_size: encoded_nsize_int = undefined;
    var encoded_size_as_bytes: []u8 = asBytes(&encoded_size);
    encode(asBytes(&size)[0..nsize_int_byte_count], encoded_size_as_bytes[0..encoded_nsize_int_byte_count]);
    for (0..encoded_nsize_int_byte_count) |i| {
        encoded_size_as_bytes[i] |= 1 << 7;
    }
    return encoded_size;
}

pub fn decodeSize(size: encoded_nsize_int) nsize_int {
    var mutsize = size;
    var encoded_size_as_bytes: []u8 = asBytes(&mutsize);
    for (0..encoded_nsize_int_byte_count) |i| {
        encoded_size_as_bytes[i] &= ~(@as(u8, 1) << 7);
    }
    var decoded_size: nsize_int = undefined;
    decode(encoded_size_as_bytes[0..encoded_nsize_int_byte_count], asBytes(&decoded_size)[0..nsize_int_byte_count]);
    return decoded_size;
}

pub fn encodeWithSize(data: []const u8, encoded: []u8) void {
    encode(data, encoded[encoded_nsize_int_byte_count..]);
    const encoded_size: encoded_nsize_int = encodeSize(@intCast(data.len));
    const encoded_size_as_bytes: []const u8 = asBytes(&encoded_size);

    for (0..encoded_nsize_int_byte_count) |i| {
        encoded[i] = encoded_size_as_bytes[i];
    }
}
